detox/src/realms/DetoxPrimaryContext.js

```diff
@@ -227,11 +237,18 @@ class DetoxPrimaryContext extends DetoxContext {
       return;
     }
 
-    if (this[_deviceAllocator]) {
-      this[_deviceAllocator].emergencyCleanup();
-      this[_deviceAllocator] = null;
+    for (const key of Object.keys(this[_deviceAllocators])) {
+      const deviceAllocator = this[_deviceAllocators][key];
+      delete this[_deviceAllocators][key];
+      try {
+        deviceAllocator.emergencyCleanup();
+      } catch (err) {
+        this[symbols.logger].error({ cat: 'device', err }, `Failed to emergency cleanup device allocator for ${key}`);
+      }
```

Comment:
Could you try to avoid the (subtle) iteration-code duplication?

Reply:
Unfortunately not. At least, without obfuscating the code further. The goal is to make sure that the function has 0% `async` statements inside, and the duplication here is the lesser of two evils.